src/Processors/QueryPlan/DistributedCreateLocalPlan.cpp

```diff
@@ -0,0 +1,76 @@
+#include <Processors/QueryPlan/DistributedCreateLocalPlan.h>
+#include <Common/checkStackSize.h>
+#include <Processors/QueryPlan/ExpressionStep.h>
+#include <Interpreters/ActionsDAG.h>
+#include <Interpreters/InterpreterSelectQuery.h>
+
+namespace DB
+{
+
+namespace
+{
+
+void addConvertingActions(QueryPlan & plan, const Block & header)
+{
+    if (blocksHaveEqualStructure(plan.getCurrentDataStream().header, header))
+        return;
+
+    auto get_converting_dag = [](const Block & block_, const Block & header_)
+    {
+        /// Convert header structure to expected.
+        /// Also we ignore constants from result and replace it with constants from header.
+        /// It is needed for functions like `now64()` or `randConstant()` because their values may be different.
+        return ActionsDAG::makeConvertingActions(
+            block_.getColumnsWithTypeAndName(),
+            header_.getColumnsWithTypeAndName(),
+            ActionsDAG::MatchColumnsMode::Name,
+            true);
+    };
+
+    auto convert_actions_dag = get_converting_dag(plan.getCurrentDataStream().header, header);
+    auto converting = std::make_unique<ExpressionStep>(plan.getCurrentDataStream(), convert_actions_dag);
+    plan.addStep(std::move(converting));
+}
+
+}
+
+std::unique_ptr<QueryPlan> createLocalPlan(
+    const ASTPtr & query_ast,
+    const Block & header,
+    ContextPtr context,
+    QueryProcessingStage::Enum processed_stage,
+    UInt32 shard_num,
+    UInt32 shard_count,
+    std::shared_ptr<ParallelReplicasReadingCoordinator> coordinator)
+{
+    checkStackSize();
+
+    auto query_plan = std::make_unique<QueryPlan>();
+    auto mutable_context = const_pointer_cast<Context>(context);
```

Comment:
That does not seem to be safe to me.
Note, here inside InterpreterSelectQuery we copy context:
https://github.com/ClickHouse/ClickHouse/blob/bcbab2ead84b428956f186913a4282c4d411b0a6/src/Interpreters/IInterpreterUnionOrSelectQuery.h#L16 (and that's why shard_num and shard_count are specified over SelectQueryOptions).

Maybe it would be better to:
* create InterpreterSelectQuery first
* add a method inside InterpreterSelectQuery which setMergeTreeReadTaskCallback
* buildQueryPlan